### PR TITLE
[comgr] Build ASM kernels

### PR DESCRIPTION
**Pull request by atamazov**
- Added support for assembling kernels via comgr.
- Added support for binary kernels.
- Finally disabled spare build path (via temp file) when comgr is used.
- Refactored workaround for the comgr MT issue.

## Testing

No correctness problems. No GPU time and Wall time regressions. Some (~1.8X) Aux Wall time improvements. Logs & csv files can be provided by request.

### CODE REVIEW

**Comment by daniellowell** [comment]
@atamazov is still considered a "draft"?

**Comment by atamazov** [comment]
I need to fix the driver to test this rigorously.

**Comment by atamazov** [comment]
Not a draft anymore. Ready for review. Under testing right now.

File: src/comgr.cpp [hunk 1]
```diff
@@ -706,5 +718,46 @@ void BuildOcl(const std::string& name,
...
+    }
+    catch(ComgrError& ex)
+    {
+        MIOPEN_LOG_E("comgr status = " << GetStatusText(ex.status));
+        if(!ex.text.empty())
+            MIOPEN_LOG_W(ex.text);
+    }
+}
+
 } // namespace comgr
```
Comment (JehandadKhan) [inline src/comgr.cpp:762]: The comment needs to be updated

Comment (atamazov) [reply src/comgr.cpp:762]: Which one? This curly brace closes `miopen::comgr`

**Comment by atamazov** [comment]
Testing done.

**Comment by atamazov** [comment]
CI checks passed, please review. Thanks.

File: src/comgr.cpp [hunk 2]
```diff
@@ -94,6 +94,18 @@ namespace compiler {
...
 #define OCL_EARLY_INLINE 1
 
+namespace gcnasm {
+
+static void RemoveOptionsUnwanted(OptionList& list)
+{
+    list.erase(remove_if(list.begin(),
+                         list.end(),
+                         [&](const auto& option) { return StartsWith(option, "-mcpu="); }),
+               list.end());
```
Comment (DrizztDoUrden) [inline src/comgr.cpp:104]:
```suggestion
    auto is_mcpu = [&](const auto& option) { return StartsWith(option, "-mcpu=");
    auto remove_it = remove_if(list.begin(), list.end(), is_mcpu);
    list.erase(remove_it, list.end());
```

This stuff is usually easier to understand when split.

**Pull request merged**